### ISSUE TITLE
OSCD results

### ISSUE THREAD

**Issue by robmarkcole** [wangle53/CSA-CDGAN#6]
Hi, In `make_dataset.py` you have `class OSCD_TRAIN` indicating you trained on the OSCD dataset, however the results are not documented. Could you comment on what you achieved?
Thank you

**Comment by wangle53** [comment]
Thanks for your interest. At the begining, I indeeded plan to use OSCD dataset, but give up later. If you are interest in OSCD dataset, please you can experiment CSA-CDGAN on it.

**Comment by robmarkcole** [comment]
I've had some encouraging results on 3 channel OSCD, but it is a small dataset which I suspect may be a limiting factor in achieving great results

**Comment by wangle53** [comment]
Maybe, you can try decrease the model parameters, such as reduce the number of channels of CNN layer.

**Comment by robmarkcole** [comment]
The NDF & NGF const?

**Comment by wangle53** [comment]
Yes this's right. To further reduce the model parameters, you maybe revise the channel number of all Conv layers.

**Comment by robmarkcole** [comment]
I gave it a try, it definitely affected the output but not sure what to make of it

<img width="1224" alt="image" src="https://github.com/wangle53/CSA-CDGAN/assets/11855322/fbd93e9f-f6df-455b-8c9c-0dfda88315dc">

**Comment by wangle53** [comment]
I think the following reasons lead to the poor effect. 1. the dataset is not large engough, the model training is not sufficent. 2. The model is underfitting because of the small model parameters. 3. The training epoch is not engough. Therefore, I recommend that you maybe increase training epoch, or retify channel numbers of conv layer, or improve learning rate.

**Comment by robmarkcole** [comment]
Increasing epochs from 100 to 1000 improved the result:
<img width="1427" alt="image" src="https://github.com/wangle53/CSA-CDGAN/assets/11855322/842745ce-d12c-4d95-ba6f-1fada12bf1ff">

**Comment by wangle53** [comment]
It seems the result is not very fine.

**Comment by robmarkcole** [comment]
The losses do not look correct - they should reach equilibrium right? 

<img width="1194" alt="image" src="https://github.com/wangle53/CSA-CDGAN/assets/11855322/3af8309c-f2f7-4766-b1ad-2fcbbef16a50">

**Comment by wangle53** [comment]
Normally, both losses should be constantly decreasing.

**Comment by robmarkcole** [comment]
I'm using lightning rather than your code implementation - perhaps you can spot any errors? 

```python3
class GanModel(LightningModule):
    def __init__(
        self,
        input_image_size: int = 256,
        latent_vector_size: int = 100,
        in_channels: int = 3,
        mask_channels: int = 1,
        use_pretrained: bool = False,
        conv_dimsize: int = 64,
        learning_rate: float = 0.0001, # 0.0002
    ):
        
        super().__init__()
        self.automatic_optimization = False # Manually manages the optimization loop, required for multiple optimizers.
        self.save_hyperparameters()
        self.NGF = conv_dimsize        # the dimension size of the first convolutional of the generator
        self.NDF = conv_dimsize        # the dimension size of the first convolutional of the discriminator
        self.EXTRALAYERS = 3 # add extral layers for the generator and discriminator
        self.G_WEIGHT = 200    # loss weight
        self.D_WEIGHT = 1        # loss weight
        self.LR_STEP_SIZE = 50
        self.GAMMA = 0.1

        self.netg = NetG(input_image_size, in_channels*2, latent_vector_size, self.NGF, self.EXTRALAYERS)
        self.netd = NetD(input_image_size, mask_channels, 1, self.NDF, self.EXTRALAYERS)
        self.netg.apply(weights_init) # if you want to re-initialize the weights
        self.netd.apply(weights_init)

        if use_pretrained:
            # Select which model weights to begin fom
            # self.model_path = '/teamspace/studios/this_studio/CDGAN/weights-official-repo/levir-cd/netG.pth'
            self.model_path = '/teamspace/studios/this_studio/CDGAN/weights-official-repo/cdd/netG.pth'
            # model_path = '/teamspace/studios/this_studio/CDGAN/weights-official-repo/whu-cd/netG.pth'
            pretrained_dict = torch.load(self.model_path, map_location=torch.device(self.device))['model_state_dict']
            self.netg.load_state_dict(pretrained_dict, False)

        self.l_adv = l2_loss
        self.l_con = nn.L1Loss()
        self.l_bce = nn.BCELoss()

        self.train_metrics = []
        self.valid_metrics = []
        self.valid_confmat = ConfusionMatrix(task="binary", num_classes=2)
        self.test_metrics = []
        self.test_confmat = ConfusionMatrix(task="binary", num_classes=2)

    def forward(self, image):
        return self.netg(image)

    def shared_step(self, batch, stage):
        opt_g, opt_d = self.optimizers() if stage == "train" else (None, None)
        image, gt = batch["image"], batch["mask"]
        # gt = gt[:,0,:,:].unsqueeze(1) # ensure reading only first dimension
        real_label = torch.ones(size=(image.shape[0],), dtype=torch.float32, device=self.device)
        fake_label = torch.zeros(size=(image.shape[0],), dtype=torch.float32, device=self.device)

        fake = self(image)
        pred_real = self.netd(gt)
        pred_fake = self.netd(fake).detach()

        err_d_fake = self.l_bce(pred_fake, fake_label)
        err_g = self.l_con(fake, gt)
        err_g_total = self.G_WEIGHT*err_g + self.D_WEIGHT*err_d_fake

        err_d_real = self.l_bce(pred_real, real_label)
        err_d_total = (err_d_real + err_d_fake) * 0.5

        metrics_dict = {
            f"{stage}_err_d_real": err_d_real,
            f"{stage}_err_d_fake": err_d_fake,
            f"{stage}_err_d_total": err_d_total,
            f"{stage}_err_g": err_g,
            f"{stage}_err_g_total": err_g_total
        }
        
        # the backward pass and optimizer steps are skipped for validation and test stages to save computation and prevent model updates.
        if stage == "train":
            # Backward and optimize for generator
            self.manual_backward(err_g_total)
            opt_g.step()
            opt_g.zero_grad()

            # Backward and optimize for discriminator
            self.manual_backward(err_d_total)
            opt_d.step()
            opt_d.zero_grad()
            self.train_metrics.append(metrics_dict)
        elif stage == "val":
            self.valid_metrics.append(metrics_dict)
        else:  # test
            self.test_metrics.append(metrics_dict)

        if stage in ["val", "test"]:
            gt = torch.argmax(batch["mask"], dim=1)
            fake = torch.argmax(self(batch["image"]), dim=1)
            confmat = self.valid_confmat if stage == "val" else self.test_confmat
            confmat.update(preds=fake, target=gt)

        return err_g_total, err_d_total  # Return both losses

    def training_step(self, batch, batch_idx):
        self.shared_step(batch, "train")

    def validation_step(self, batch, batch_idx):
        self.shared_step(batch, "val")

    def test_step(self, batch, batch_idx):
        self.shared_step(batch, "test")

    def on_train_epoch_end(self):
        if self.train_metrics:
            avg_train_metrics = {k: torch.stack([x[k] for x in self.train_metrics]).mean() for k in self.train_metrics[0]}
            self.log_dict(avg_train_metrics, prog_bar=True)
            self.train_metrics.clear()

    def on_validation_epoch_end(self):
        if self.valid_metrics:
            avg_valid_metrics = {k: torch.stack([x[k] for x in self.valid_metrics]).mean() for k in self.valid_metrics[0]}
            self.log_dict(avg_valid_metrics, prog_bar=True)
            self.valid_metrics.clear()
        # Handle confusion matrix
        # fig_valid, ax_ = self.valid_confmat.plot()
        # self.logger.experiment.log_figure(
        #     run_id=self.logger.run_id,
        #     figure=fig_valid,
        #     artifact_file=f"val_epoch_{self.current_epoch}.png",
        # )
        # self.valid_confmat.reset()

    def on_test_epoch_end(self):
        if self.test_metrics:
            avg_test_metrics = {k: torch.stack([x[k] for x in self.test_metrics]).mean() for k in self.test_metrics[0]}
            self.log_dict(avg_test_metrics, prog_bar=True)
            self.test_metrics.clear()
        # Handle confusion matrix
        # fig_test, ax_ = self.test_confmat.plot()
        # self.logger.experiment.log_figure(
        #     run_id=self.logger.run_id,
        #     figure=fig_test,
        #     artifact_file=f"test_epoch_{self.current_epoch}.png",
        # )

    def configure_optimizers(self):
        opt_g = torch.optim.Adam(self.netg.parameters(), lr=self.hparams.learning_rate, betas=(0.5, 0.999))
        opt_d = torch.optim.Adam(self.netd.parameters(), lr=self.hparams.learning_rate, betas=(0.5, 0.999))
        return [opt_g, opt_d], []

    # def configure_optimizers(self):
    #     opt_g = torch.optim.Adam(self.netg.parameters(), lr=self.hparams.learning_rate, betas=(0.5, 0.999))
    #     opt_d = torch.optim.Adam(self.netd.parameters(), lr=self.hparams.learning_rate, betas=(0.5, 0.999))
        
    #     scheduler_g = StepLR(opt_g, step_size=self.LR_STEP_SIZE, gamma=self.GAMMA)
    #     scheduler_d = StepLR(opt_d, step_size=self.LR_STEP_SIZE, gamma=self.GAMMA)

    #     return {
    #         'optimizer': [opt_g, opt_d],
    #         'lr_scheduler': [
    #             {'scheduler': scheduler_g, 'interval': 'epoch', 'frequency': 1},
    #             {'scheduler': scheduler_d, 'interval': 'epoch', 'frequency': 1}
    #         ]
    #     }
```

**Comment by wangle53** [comment]
For OSCD，I think the reason is that the loss weight of discriminator is too small，since the learning rate is small too，the discriminator seems doesn't be  updated effectively. Maybe you can     increase self.D_WEIGHT.

**Comment by robmarkcole** [comment]
Increased from 1 -> 10

<img width="1424" alt="image" src="https://github.com/wangle53/CSA-CDGAN/assets/11855322/3a648343-5db3-484e-9ae6-a34689072807">

And to 100

<img width="1427" alt="image" src="https://github.com/wangle53/CSA-CDGAN/assets/11855322/2acd1cdb-e881-4d8a-b74b-e1a4822597f2">

**Comment by wangle53** [comment]
It seems the model doesn't perform well on OSCD datadet.
